// @elizaos/cli/dist/chunk-I77ZRNYO.js
import {
  createAgent,
  createPlugin,
  createProject,
  createTEEProject,
  hasEmbeddingSupport,
  selectAIModel,
  selectDatabase,
  selectEmbeddingModel,
  validateCreateOptions,
  validateProjectName
} from "./chunk-2CUIHNPL.js";
import {
  displayBanner,
  handleError
} from "./chunk-2ALAPQLV.js";

// src/commands/create/index.ts
import { Command } from "commander";
import * as clack from "@clack/prompts";
import colors from "yoctocolors";
import { logger } from "@elizaos/core";
function formatProjectType(type) {
  return type === "tee" ? "TEE Project" : type.charAt(0).toUpperCase() + type.slice(1);
}
var create = new Command("create").description("Create a new ElizaOS project, plugin, agent, or TEE project").argument("[name]", "name of the project/plugin/agent to create").option("--yes, -y", "skip prompts and use defaults").option("--type <type>", "type of project to create (project, plugin, agent, tee)", "project").action(async (name, opts) => {
  let projectType;
  try {
    if (process.env.ELIZA_NONINTERACTIVE === "1" || process.env.ELIZA_NONINTERACTIVE === "true" || process.argv.includes("-y") || process.argv.includes("--yes")) {
      if (opts) {
        opts.yes = true;
      } else {
        opts = { yes: true };
      }
    }
    const options = validateCreateOptions(opts || {});
    const isNonInteractive = options.yes;
    if (!isNonInteractive) {
      await displayBanner();
    }
    projectType = options.type;
    let projectName = name;
    if (!projectName) {
      if (!isNonInteractive) {
        const selectedType = await clack.select({
          message: "What would you like to create?",
          options: [
            {
              label: "Project - Full ElizaOS application",
              value: "project",
              hint: "Complete project with runtime, agents, and all features"
            },
            {
              label: "Plugin - Custom ElizaOS plugin",
              value: "plugin",
              hint: "Extend ElizaOS functionality with custom plugins"
            },
            {
              label: "Agent - Character definition file",
              value: "agent",
              hint: "Create a new agent character file"
            },
            {
              label: "TEE Project - Trusted Execution Environment project",
              value: "tee",
              hint: "Secure computing environment for privacy-focused applications"
            }
          ],
          initialValue: "project"
        });
        if (clack.isCancel(selectedType)) {
          clack.cancel("Operation cancelled.");
          process.exit(0);
        }
        projectType = selectedType;
      }
      if (!isNonInteractive) {
        const introType = formatProjectType(projectType);
        clack.intro(colors.inverse(` Creating ElizaOS ${introType} `));
      }
      if (!isNonInteractive) {
        const nameInput = await clack.text({
          message: `What is the name of your ${projectType}?`,
          placeholder: `my-${projectType}`,
          validate: (value) => {
            if (!value) return "Name is required";
            if (projectType === "agent") {
              return value.length > 0 ? void 0 : "Agent name cannot be empty";
            }
            const validation = validateProjectName(value);
            return validation.isValid ? void 0 : validation.error;
          }
        });
        if (clack.isCancel(nameInput)) {
          clack.cancel("Operation cancelled.");
          process.exit(0);
        }
        projectName = nameInput;
      } else {
        throw new Error(`Project name is required. Usage: elizaos create [name]`);
      }
    }
    if (projectType !== "agent") {
      const nameValidation = validateProjectName(projectName);
      if (!nameValidation.isValid) {
        throw new Error(nameValidation.error);
      }
    }
    if (!isNonInteractive && name) {
      const introType = formatProjectType(projectType);
      clack.intro(colors.inverse(` Creating ElizaOS ${introType} `));
    }
    switch (projectType) {
      case "plugin": {
        let pluginType = "full";
        if (!isNonInteractive) {
          const selectedPluginType = await clack.select({
            message: "What type of plugin would you like to create?",
            options: [
              {
                label: "Quick Plugin (Backend Only)",
                value: "quick",
                hint: "Simple backend-only plugin without frontend"
              },
              {
                label: "Full Plugin (with Frontend)",
                value: "full",
                hint: "Complete plugin with React frontend and API routes"
              }
            ],
            initialValue: "quick"
          });
          if (clack.isCancel(selectedPluginType)) {
            clack.cancel("Operation cancelled.");
            process.exit(0);
          }
          pluginType = selectedPluginType;
        }
        await createPlugin(projectName, process.cwd(), pluginType, isNonInteractive);
        break;
      }
      case "agent":
        await createAgent(projectName, process.cwd(), isNonInteractive);
        break;
      case "tee": {
        let database = "pglite";
        let aiModel = "local";
        let embeddingModel;
        if (!isNonInteractive) {
          database = await selectDatabase();
          aiModel = await selectAIModel();
          if (!hasEmbeddingSupport(aiModel)) {
            embeddingModel = await selectEmbeddingModel();
          }
        }
        await createTEEProject(
          projectName,
          process.cwd(),
          database,
          aiModel,
          embeddingModel,
          isNonInteractive
        );
        break;
      }
      case "project":
      default: {
        let database = "pglite";
        let aiModel = "local";
        let embeddingModel;
        if (!isNonInteractive) {
          database = await selectDatabase();
          aiModel = await selectAIModel();
          if (!hasEmbeddingSupport(aiModel)) {
            embeddingModel = await selectEmbeddingModel();
          }
        }
        await createProject(
          projectName,
          process.cwd(),
          database,
          aiModel,
          embeddingModel,
          isNonInteractive
        );
        break;
      }
    }
    if (!isNonInteractive) {
      const typeLabel = formatProjectType(projectType);
      clack.outro(colors.green(`${typeLabel} created successfully! \u{1F389}`));
    }
  } catch (error) {
    if (!opts?.yes) {
      const errorType = formatProjectType(projectType || "project");
      clack.cancel(`Failed to create ${errorType}.`);
    }
    logger.error("Create command failed:", error);
    handleError(error);
    process.exit(1);
  }
});

export {
  create
};
